### PR TITLE
Arnej/better remove strategy 

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@geirst please review
@havardpe please take a look

note: i know i need to add/fix unit test, mostly wanted to checkpoint my work.

### CODE REVIEW

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.h [hunk 1]
```diff
@@ -101,6 +101,11 @@ class HnswIndex : public NearestNeighborIndex {
     LevelArrayRef get_level_array(uint32_t docid) const;
     LinkArrayRef get_link_array(uint32_t docid, uint32_t level) const;
     void set_link_array(uint32_t docid, uint32_t level, const LinkArrayRef& links);
+    void add_link_to(uint32_t docid, uint32_t level, const LinkArrayRef& old_links, uint32_t new_link) {
+        LinkArray new_links(old_links.begin(), old_links.end());
+        new_links.push_back(new_link);
+        set_link_array(docid, level, new_links);
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index.h]: since you set the link array specifically, maybe you can get it specifically as well?

**Comment by havardpe** [comment]
Looks good, but I do not understand why adding more links now allows more links to be added later(?)

**Comment by arnej27959** [comment]
> Looks good, but I do not understand why adding more links now allows more links to be added later(?)

always filtering means adding less links now to allow more links to be added later, at least that is the intention...

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp [hunk 2]
```diff
@@ -334,6 +346,37 @@ HnswIndex::add_document(uint32_t docid)
...
+        uint32_t n_id_1 = cluster[i];
+        LinkArrayRef n_list_1 = get_link_array(n_id_1, level);
+        for (uint32_t j = i + 1; j < cluster.size(); ++j) {
+            uint32_t n_id_2 = cluster[j];
+            if (has_link_to(n_list_1, n_id_2)) continue;
+            pairs.emplace_back(n_id_1, n_id_2, calc_distance(n_id_1, n_id_2));
+        }
+    }
+    std::sort(pairs.begin(), pairs.end());
+    for (const PairDist & pair : pairs) {
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: Consider using auto.

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp [hunk 3]
```diff
@@ -334,6 +346,37 @@ HnswIndex::add_document(uint32_t docid)
...
+void
+HnswIndex::mutual_reconnect(const LinkArrayRef &cluster, uint32_t level)
+{
+    std::vector<PairDist> pairs;
+    for (uint32_t i = 0; i + 1 < cluster.size(); ++i) {
+        uint32_t n_id_1 = cluster[i];
+        LinkArrayRef n_list_1 = get_link_array(n_id_1, level);
+        for (uint32_t j = i + 1; j < cluster.size(); ++j) {
+            uint32_t n_id_2 = cluster[j];
+            if (has_link_to(n_list_1, n_id_2)) continue;
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: Prefer multi-line with brackets instead of one-line if statements.

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**